docs/proposals/28-autonomous-shoot-clusters.md

```diff
@@ -0,0 +1,498 @@
+- Sufficient CPU and RAM (to be specified more concretely later).
+- Full network connectivity between all machines in the cluster (public or private network is fine).
+- [Certain ports](https://kubernetes.io/docs/reference/networking/ports-and-protocols/) must be open.
+- [systemd](https://www.freedesktop.org/wiki/Software/systemd/), [containerd](https://containerd.io/), and its [ctr (client CLI)](https://github.com/projectatomic/containerd/blob/master/docs/cli.md/) must be pre-installed.
+
+### Commands
+
+`gardenadm` will have the following commands:
+
+- [`gardenadm discover`](#gardenadm-discover) to discover the configuration of the Gardener system the autonomous shoot cluster shall be connected to later.
+- [`gardenadm init`](#gardenadm-init) to bootstrap a first control plane node.
+- [`gardenadm token`](#gardenadm-token) to manage bootstrap and discovery tokens for `gardenadm join`.
+- [`gardenadm join`](#gardenadm-join) to bootstrap further control plane nodes, or to bootstrap worker nodes and join them to the cluster.
+- [`gardenadm connect`](#gardenadm-connect) to deploy a `gardenlet` for further cluster management.
+- [`gardenadm version`](#gardenadm-version) to print the `gardenadm` version.
+
+The general flow for bootstrapping an autonomous shoot cluster will be as follows:
+
+1. The user installs `gardenadm` on the first control plane node and runs `gardenadm discover`.
+   This will download relevant configuration from the Gardener system the autonomous shoot cluster shall be later connected to (e.g., `Controller{Registration,Deployment}`s, `CloudProfile`s, etc.).
+   If there is no existing Gardener system yet, the needed configuration must be provided manually.
+2. The downloaded configuration and a `Shoot` manifest are provided to the `gardenadm init` command.
+   This will bootstrap the first control plane node.
+3. Optionally, additional control plane nodes can be joined.
+   In order to do so, the user runs `gardenadm token generate` which generates a random bootstrap token, and then `gardenadm token create <token>` which creates it on the server.
+   Now connected to the new node that shall be joined, the user runs the `gardenadm join` command which bootstraps another control plane node.
+5. Optionally, worker nodes can be joined.
+   The flow is the same - a bootstrap token must be generated and created, and then `gardenadm join` must be executed to bootstrap the worker node.
+6. Once above steps are completed, the user runs `gardenadm connect` to deploy a `gardenlet` into the cluster which connects to an existing Gardener system for further cluster management.
+
+A more detailed description of the commands is following.
+
+#### `gardenadm discover`
+
+This command takes a kubeconfig to an existing Gardener system, and a `Shoot` manifest.
+It downloads the required `CloudProfile`, `ControllerRegistration` and `ControllerDeployment` resources and stores them locally for usage of other `gardenadm` commands.
```

Comment:
Once the `gardenlet` is deployed with `gardenadm connect`, it just reads such resources directly from the garden cluster. These resources are only needed locally for the initial bootstrapping (since there is no connection to a garden cluster yet).